File: solutions/leetcode/unique-length-3-palindromic-subsequences/main.cpp
class Solution {
   public:
    int countPalindromicSubsequence(string s) {
        array<vector<int>, 26> charIndices = {};
        for (int i = 0; i < s.size(); ++i) {
            charIndices[s[i] - 'a'].push_back(i);
        }

        int result = 0;
        for (const vector<int>& indices : charIndices) {
            if (indices.size() < 2) continue;
            unordered_set<char> chars;
            for (int i = indices.front() + 1; i < indices.back(); ++i) {
                chars.insert(s[i]);
            }
            result += chars.size();
        }

        return result;
    }
};

File: solutions/leetcode/unique-length-3-palindromic-subsequences/main.py
class Solution:

    def countPalindromicSubsequence(self, s: str) -> int:
        char_indices = defaultdict(lambda: [])
        for i, c in enumerate(s):
            char_indices[c].append(i)

        result = 0

        for indices in char_indices.values():
            if len(indices) == 1:
                continue
            result += len(set(s[indices[0] + 1:indices[-1]]))

        return result
